### PR TITLE
Log nonce

### PR DESCRIPTION
**Pull request by iljabvh**
The branch resolves issue https://github.com/hyperledger-labs/perun-eth-backend/issues/41

### CODE REVIEW

File: channel/contractbackend.go [hunk 1]
```diff
@@ -62,25 +62,116 @@ type Transactor interface {
...
 	nonceMtx          *sync.Mutex
+}
+
+// GlobalNoncer implements a global nonce counter that is shared across different contract backends.
+type GlobalNoncer struct {
 	expectedNextNonce map[common.Address]uint64
-	txFinalityDepth   uint64
-	chainID           ChainID
+	nonceMtx          map[common.Address]*sync.Mutex
+}
```
Comment (ndzik) [inline channel/contractbackend.go:85]: The `GlobalNoncer` requires a mutex to guarantee only a single nonce is created for each `address` entry.
Furthermore I would rename the struct to `SharedNoncer`. But I am honestly a little bit surprised by the definition.
`LocalNoncer` and `GlobalNoncer` effectively do the same: Guarding access to the `Nonce` for a specific `common.Address`. The only difference is, that the `LocalNoncer` is more restrictive in that it only allows updates to one nonce entry at a time.
The `GlobalNoncer` on the other hand allows updates to multiple nonces at the same time within the same map. Thinking about it even more this might create another issue where two routines try to write to the same map at once, which would require another mutex anyway, making `GlobalNoncer` just a more complicated and convoluted implementation of the `LocalNoncer`.

Long story short: I would get rid of the `GlobalNoncer`, rename `LocalNoncer` to `SharedNoncer` and accept the simpler implementation of a single mutex guarding all nonces.

File: channel/contractbackend.go [hunk 2]
```diff
@@ -62,25 +62,116 @@ type Transactor interface {
...
+
+func (d GlobalNoncer) setNonce(chainID ChainID, sender common.Address, nonce uint64) (uint64, error) {
```
Comment (ndzik) [inline channel/contractbackend.go:87]:
```suggestion
func (d *GlobalNoncer) setNonce(chainID ChainID, sender common.Address, nonce uint64) (uint64, error) {
```

Should be a pointer receiver.

File: channel/contractbackend.go [hunk 3]
```diff
@@ -62,25 +62,116 @@ type Transactor interface {
...
+		nonce = expectedNextNonce
+	}
+
+	// Update local expectation.
+	d.expectedNextNonce[sender] = nonce + 1
+
+	return nonce, nil
+}
+
+func (d LocalNoncer) setNonce(chainID ChainID, sender common.Address, nonce uint64) (uint64, error) {
```
Comment (ndzik) [inline channel/contractbackend.go:111]:
```suggestion
func (d *LocalNoncer) setNonce(chainID ChainID, sender common.Address, nonce uint64) (uint64, error) {
```

Applying my comment from my small essay regarding the `GlobalNoncer` would require this to be a pointer receiver as well.

**Review by ndzik — CHANGES_REQUESTED** [review]

**Comment by iljabvh** [comment]
Issue [#41](https://github.com/hyperledger-labs/perun-eth-backend/issues/41) fixed by [#43](https://github.com/hyperledger-labs/perun-eth-backend/pull/43).
